BinaryEncoding.md

```diff
@@ -317,18 +317,20 @@ It is legal to have several entries with the same type.
 | Name | Opcode | Immediate | Description |
 | ---- | ---- | ---- | ---- |
 | `nop` | `0x00` | | no operation |
-| `block` | `0x01` | count = `varuint32` | a sequence of expressions, the last of which yields a value |
-| `loop` | `0x02` | count = `varuint32` | a block which can also form control flow loops |
-| `if` | `0x03` | | high-level one-armed if |
-| `if_else` | `0x04` | | high-level two-armed if |
+| `block` | `0x01` |  | begin a sequence of expressions, the last of which yields a value |
+| `loop` | `0x02` |  | begin a block which can also form control flow loops |
+| `if` | `0x03` | | begin if expression |
+| `else` | `0x04` | | begin else expression of if |
 | `select` | `0x05` | | select one of two values based on condition |
 | `br` | `0x06` | relative_depth = `varuint32` | break that targets a outer nested block |
 | `br_if` | `0x07` | relative_depth = `varuint32` | conditional break that targets a outer nested block |
 | `br_table` | `0x08` | see below | branch table control flow construct |
-| `return` | `0x14` | | return zero or one value from this function |
-| `unreachable` | `0x15` | | trap immediately |
+| `return` | `0x09` | | return zero or one value from this function |
+| `unreachable` | `0x0a` | | trap immediately |
+| `end` | `0x0f` | | end a block, loop, or if |
```

Comment:
On Tue, Mar 29, 2016 at 4:33 PM, Dan Gohman <notifications@github.com>
wrote:

> In BinaryEncoding.md
> <https://github.com/WebAssembly/design/pull/628#discussion_r57733287>:
>
> >  | `select` | `0x05` | | select one of two values based on condition |
> >  | `br` | `0x06` | relative_depth = `varuint32` | break that targets a outer nested block |
> >  | `br_if` | `0x07` | relative_depth = `varuint32` | conditional break that targets a outer nested block |
> >  | `br_table` | `0x08` | see below | branch table control flow construct |
> > -| `return` | `0x14` | | return zero or one value from this function |
> > -| `unreachable` | `0x15` | | trap immediately |
> > +| `return` | `0x09` | | return zero or one value from this function |
> > +| `unreachable` | `0x0a` | | trap immediately |
> > +| `end` | `0x0f` | | end a block, loop, or if |
>
> In my postorder design, if has its own endif opcode. block and loop ends
> have an arity immediate, while in if+else+endif the arity immediate goes
> on the if to avoid being redundant between the else and the endif. Does
> your design omit arity immediates on end markers, and if so, how do you
> determine what to leave on the stack after a block exit?
>
Yes, I omit the arities, so that when you fall off the end of the block,
you can have at most one return value (i.e. the last value on the stack). I
think that's sufficient; if you want to use a multi-value block in the
future, you have to use a br with arity. (And I'm still not convinced that
multi-arity blocks and ifs will be a thing in the future).



> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/WebAssembly/design/pull/628/files/2766cd2fa08155d85e3e6197cacdf4921a3ff253#r57733287>
>